service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/C21OrderController.java

```diff
@@ -0,0 +1,113 @@
+        this.service = service;
+        this.validateGroupService = validateGroupService;
+        this.docmosisDocumentGeneratorService = docmosisDocumentGeneratorService;
+        this.uploadDocumentService = uploadDocumentService;
+    }
+
+    @PostMapping("/about-to-start")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .errors(validateGroupService.validateGroup(caseData, C21CaseOrderGroup.class))
+            .build();
+    }
+
+    @PostMapping("/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleMidEvent(
+        @RequestHeader(value = "authorization") String authorization,
+        @RequestHeader(value = "user-id") String userId,
+        @RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        Document c21Document = getDocument(authorization, userId, caseData);
+
+        caseDetails.getData().put("c21Order", service.addDocumentToC21(caseData.getC21Order(), c21Document));
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .build();
+    }
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(
+        @RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = mapper.convertValue(caseDetails.getData(), CaseData.class);
+
+        List<Element<C21Order>> c21Orders = caseData.getC21Orders();
+        c21Orders.add(Element.<C21Order>builder()
+            .id(randomUUID())
+            .value(service.addCustomValuesToC21Order(caseData.getC21Order(), caseData.getJudgeAndLegalAdvisor()))
+            .build());
+
+        caseDetails.getData().put("c21Orders", c21Orders);
+        caseDetails.getData().remove("c21Order");
+        caseDetails.getData().remove("judgeAndLegalAdvisor");
+
+        return AboutToStartOrSubmitCallbackResponse.builder()
+            .data(caseDetails.getData())
+            .build();
+    }
+
+    public Document getDocument(@RequestHeader("authorization") String authorization,
+                                @RequestHeader("user-id") String userId,
+                                CaseData caseData) {
+        DocmosisDocument document = docmosisDocumentGeneratorService.generateDocmosisDocument(
+            service.getC21OrderTemplateData(caseData), C21);
+        String index = (caseData.getC21Orders() != null) ? Integer.toString(caseData.getC21Orders().size() + 1) : "1";
```

Comment:
the logic on line 108 worries me. Currently isn't tested. Can we extract?

Reply:
Extracted and tested